File: packages/sdk-python/regressa/anthropic.py
from __future__ import annotations

import time
from typing import Any

from .client import Regressa, is_timeout
from .openai import _ms, _pop_opts, _StreamProxy
from .template import infer_template


def wrap_anthropic(client: Any, regressa: Regressa, **defaults: Any) -> Any:
    """Instrument an `anthropic.Anthropic` client's messages.create (sync, incl. stream=True)."""
    original = client.messages.create

    def create(*args: Any, **kwargs: Any) -> Any:
        opts = _pop_opts(kwargs, defaults)
        messages = list(kwargs.get("messages", []))
        system = kwargs.get("system")
        input_messages = ([{"role": "system", "content": system}] + messages) if system else messages
        base: dict[str, Any] = {
            "model": kwargs.get("model", "unknown"),
            "provider": "anthropic",
            "input_messages": input_messages,
            "prompt_template": opts.get("prompt_template") or infer_template(messages, system),
            "trace_group_id": opts.get("trace_group_id"),
            "metadata": opts.get("metadata"),
        }
        started = time.perf_counter()
        try:
            res = original(*args, **kwargs)
        except Exception as exc:
            regressa.trace(
                **base,
                latency_ms=_ms(started),
                status="timeout" if is_timeout(exc) else "error",
                error_message=str(exc),
            )
            raise
        if kwargs.get("stream"):
            return _StreamProxy(res, regressa, base, started, _anthropic_event)
        usage = getattr(res, "usage", None)
        text = "".join(
            getattr(b, "text", "") for b in (getattr(res, "content", None) or []) if getattr(b, "type", "") == "text"
        )
        regressa.trace(
            **{**base, "model": getattr(res, "model", None) or base["model"]},
            output_text=text,
            prompt_tokens=getattr(usage, "input_tokens", None),
            completion_tokens=getattr(usage, "output_tokens", None),
            latency_ms=_ms(started),
            status="success",
        )
        return res

    client.messages.create = create
    return client


def _anthropic_event(ev: Any, state: dict[str, Any]) -> None:
    t = getattr(ev, "type", None)
    if t == "message_start":
        msg = getattr(ev, "message", None)
        state["model"] = getattr(msg, "model", None)
        state["prompt_tokens"] = getattr(getattr(msg, "usage", None), "input_tokens", None)
    elif t == "content_block_delta" and getattr(getattr(ev, "delta", None), "type", "") == "text_delta":
        state["chunks"].append(ev.delta.text)
    elif t == "message_delta":
        out = getattr(getattr(ev, "usage", None), "output_tokens", None)
        if out is not None:
            state["completion_tokens"] = out

File: packages/sdk-node/src/anthropic.ts
import type { Regressa } from "./client.js";
import type { CallOptions, Message } from "./types.js";
import { inferTemplate } from "./template.js";

interface AnthropicLike {
  messages: { create: (...args: any[]) => any };
}

type CreateParams = { model: string; messages: Message[]; system?: unknown; stream?: boolean; regressa?: CallOptions; [k: string]: unknown };

/**
 * Wrap an Anthropic client. Instruments `messages.create` (non-streaming and `stream: true`).
 *
 *   const anthropic = wrapAnthropic(new Anthropic(), regressa);
 */
export function wrapAnthropic<T extends AnthropicLike>(client: T, regressa: Regressa, defaults: CallOptions = {}): T {
  const original = client.messages.create.bind(client.messages);

  client.messages.create = (params: CreateParams, ...rest: unknown[]) => {
    const { regressa: callOpts, ...clean } = params;
    const opts: CallOptions = { ...defaults, ...callOpts };
    const started = Date.now();
    const inputMessages: Message[] = clean.system
      ? [{ role: "system", content: clean.system }, ...clean.messages]
      : clean.messages;
    const base = {
      model: clean.model,
      provider: "anthropic" as const,
      input_messages: inputMessages,
      prompt_template: opts.promptTemplate ?? inferTemplate(clean.messages, clean.system),
      trace_group_id: opts.traceGroupId,
      metadata: opts.metadata,
    };

    const promise = original(clean, ...rest);
    if (clean.stream) return wrapStream(promise, regressa, base, started);

    return promise.then(
      (res: any) => {
        regressa.trace({
          ...base,
          model: res?.model ?? clean.model,
          output_text: textOf(res?.content),
          prompt_tokens: res?.usage?.input_tokens,
          completion_tokens: res?.usage?.output_tokens,
          latency_ms: Date.now() - started,
          status: "success",
        });
        return res;
      },
      (err: unknown) => {
        regressa.trace({ ...base, latency_ms: Date.now() - started, status: "error", error_message: errMsg(err) });
        throw err;
      },
    );
  };
  return client;
}

async function wrapStream(promise: Promise<any>, regressa: Regressa, base: any, started: number) {
  const stream = await promise;
  const chunks: string[] = [];
  let inTok: number | undefined, outTok: number | undefined, model: string | undefined;
  const iter = stream[Symbol.asyncIterator].bind(stream);
  stream[Symbol.asyncIterator] = async function* () {
    try {
      for await (const ev of iter()) {
        if (ev?.type === "message_start") { inTok = ev.message?.usage?.input_tokens; model = ev.message?.model; }
        if (ev?.type === "content_block_delta" && ev.delta?.type === "text_delta") chunks.push(ev.delta.text);
        if (ev?.type === "message_delta" && ev.usage?.output_tokens != null) outTok = ev.usage.output_tokens;
        yield ev;
      }
      regressa.trace({ ...base, model: model ?? base.model, output_text: chunks.join(""), prompt_tokens: inTok, completion_tokens: outTok,
        latency_ms: Date.now() - started, status: "success" });
    } catch (err) {
      regressa.trace({ ...base, output_text: chunks.join(""), latency_ms: Date.now() - started, status: "error", error_message: errMsg(err) });
      throw err;
    }
  };
  return stream;
}

function textOf(content: unknown): string | null {
  if (!Array.isArray(content)) return null;
  return content.filter((b) => b?.type === "text").map((b) => b.text as string).join("");
}
function errMsg(err: unknown) { return err instanceof Error ? err.message : String(err); }
